drivers/epson/projector/esc_vp21.cr

```diff
@@ -0,0 +1,274 @@
+    else
+      self[:power_target] = false
+      do_send(:PWR, false, timeout: 10000, name: :power)
+      logger.debug { "-- epson Proj, requested to power off" }
+      do_send(:PWR, name: :power_state)
+    end
+  end
+
+  def power?(**options, &block)
+    options[:emit] = block unless block.nil?
+    options[:name] = :power_state
+    do_send(:PWR, **options)
+  end
+
+  #
+  # Input selection
+  #
+  INPUTS = {
+    Inputs::HDMI    => 0x30,
+    Inputs::HDBaseT => 0x80,
+  }
+  INPUT_LOOKUP = INPUTS.invert
+
+  def switch_to(input : Inputs)
+    do_send(:SOURCE, INPUT_LOOKUP[input], name: :inpt_source)
+    do_send(:SOURCE, name: :inpt_query)
+
+    logger.debug { "-- epson LCD, requested to switch to: #{input}" }
+    self[:input] = input # for a responsive UI
+    self[:MUTE] = false
+  end
+
+  #
+  # Volume commands are sent using the inpt command
+  #
+  def volume(vol : Int32, **options)
+    vol = 0 if vol < 0
+    vol = 255 if vol > 255
+
+    # Seems to only return ":" for this command
+    self[:volume] = vol
+    self[:unmute_volume] = vol if vol > 0 # Store the "pre mute" volume, so it can be restored on unmute
+    do_send(:VOL, vol, **options)
+  end
+
+  # Mutes audio + video
+  def mute(
+    state : Bool = true,
+    index : Int32 | String = 0,
+    layer : MuteLayer = MuteLayer::AudioVideo
+  )
+    logger.debug { "-- epson Proj, requested mute state: #{state}" }
+
+    do_send(:MUTE, state, name: :video_mute) # Audio + Video
+    do_send(:MUTE)                           # request status
+  end
+
+  def unmute(index : Int32 | String = 0, layer : MuteLayer = MuteLayer::AudioVideo)
+    mute false, index, layer
+  end
+
+  # Audio mute
+  def mute_audio(state : Bool = true, index : Int32 | String = 0)
+    mute state, index, MuteLayer::Audio
+
+    val = state ? 0 : self[:unmute_volume]
+    volume(val)
+  end
+
+  def unmute_audio(index : Int32 | String = 0)
+    mute_audio(false, index)
+  end
+
+  def input?
+    do_send(:SOURCE, name: :inpt_query, priority: 0)
+  end
+
+  ERRORS = {
+     0 => "00: no error",
+     1 => "01: fan error",
+     3 => "03: lamp failure at power on",
+     4 => "04: high internal temperature",
+     6 => "06: lamp error",
+     7 => "07: lamp cover door open",
+     8 => "08: cinema filter error",
+     9 => "09: capacitor is disconnected",
+    10 => "0A: auto iris error",
+    11 => "0B: subsystem error",
+    12 => "0C: low air flow error",
+    13 => "0D: air flow sensor error",
+    14 => "0E: ballast power supply error",
+    15 => "0F: shutter error",
+    16 => "10: peltiert cooling error",
+    17 => "11: pump cooling error",
+    18 => "12: static iris error",
+    19 => "13: power supply unit error",
+    20 => "14: exhaust shutter error",
+    21 => "15: obstacle detection error",
+    22 => "16: IF board discernment error",
+  }
+
+  #
+  # epson Response code
+  #
+  def received(data, resolve, command) # Data is default received as a string
```

Comment:
This method signature looks a bit off - the received callback is passed the data and a task object.